File: five_sensor_reader/SteeringController.cpp
/*
 * SteeringController.cpp
 *
 * ステアリング制御クラス（実装）
 * Follow the Gap + PD制御 + 壁回避補正
 *
 * 設計思想:
 * - GapFinderが検出したギャップ中心方向へステアリング（攻め）
 * - WallAvoiderが壁に近すぎる時に逃げる補正を追加（守り）
 * - P項: 目標角度に比例したステアリング
 * - D項: 角度変化率に比例した予測制御（高速時のオーバーシュート抑制）
 * - 調整パラメータは STEERING_KP, STEERING_KD, WALL_AVOIDANCE_GAIN
 */

#include "SteeringController.h"

#include "SensorReader.h"

SteeringController::SteeringController() : _lastTargetAngle(0.0), _wallAvoider() {}

void SteeringController::begin() {
    _lastTargetAngle = 0.0;
}

float SteeringController::calculate(const GapResult& gap, const SensorData* sensorData) {
    // 直進モード: 正面センサー（インデックス2、0度）が閾値以上空いていれば直進
    // ただし壁回避補正は適用する
    if (sensorData[2].valid &&
        sensorData[2].distance >= STRAIGHT_MODE_THRESHOLD) {
        float wall_correction = _wallAvoider.calculate(sensorData);
        _lastTargetAngle = 0.0;
        return constrain(wall_correction, -MAX_STEERING_ANGLE, MAX_STEERING_ANGLE);
    }

    // P項: 目標角度に比例
    float p_term = STEERING_KP * gap.target_angle;

    // D項: 角度変化率に比例（急変動を抑制）
    float angle_change = gap.target_angle - _lastTargetAngle;
    float d_term = -STEERING_KD * angle_change;  // 負のフィードバックで急変動を抑制

    // 壁回避補正: 壁が近すぎる時に逃げる
    float wall_correction = _wallAvoider.calculate(sensorData);

    // 目標角度を保存（次回のD項計算用）
    _lastTargetAngle = gap.target_angle;

    // PD制御 + 壁回避補正: ステアリング角 = P項 + D項 + 壁回避
    float steering = p_term + d_term + wall_correction;

    // 最大操舵角でクランプ
    steering = constrain(steering, -MAX_STEERING_ANGLE, MAX_STEERING_ANGLE);

    return steering;
}

void SteeringController::reset() { _lastTargetAngle = 0.0; }

File: src/clases/SteeringController.py
# Steering_controller.py
import math
import Config # <--- Importamos la configuración centralizada

class SteeringController:
    def __init__(self):
        # Pure Pursuit no tiene estado interno (stateless); el constructor está vacío.
        pass

    def begin(self):
        # Incluido por simetría con los otros módulos del sistema de Arduino
        pass

    def calculate(self, gap, sensor_data):
        """
        Calcula el ángulo de dirección (grados) usando Pure Pursuit.
        gap: Instancia de GapResult (contiene target_angle)
        sensor_data: Lista de objetos SensorData del SensorReader
        """
        alpha_deg = gap.target_angle

        # Restamos el offset a la distancia del sensor frontal para obtener Ld.
        # Equivalente al operador ternario ( ? : ) de C++
        front_sensor = sensor_data[Config.FRONT_SENSOR_INDEX]
        
        if front_sensor.valid:
            Ld_mm = front_sensor.distance - Config.LOOKAHEAD_OFFSET_MM
        else:
            Ld_mm = 1000.0  # Fallback de seguridad si el sensor frontal no es válido

        # Evitar división por cero o valores geométricos extremos/absurdos
        if Ld_mm < 50.0:
            Ld_mm = 50.0

        alpha_rad = alpha_deg * Config.DEG_TO_RAD

        # Fórmula matemática de Pure Pursuit: delta = atan2(2 * L * sin(alpha), Ld)
        steering_rad = math.atan2(
            2.0 * Config.WHEELBASE_MM * math.sin(alpha_rad), 
            Ld_mm
        )

        steering_deg = steering_rad * Config.RAD_TO_DEG

        # Clamp al límite mecánico del servo usando max() y min() (Equivalente al constrain)
        steering_deg = max(-Config.MAX_STEERING_ANGLE, min(Config.MAX_STEERING_ANGLE, steering_deg))

        return steering_deg
